### ISSUE TITLE
Tailscale authentication

### ISSUE THREAD

**Issue by mholt** [caddyserver/caddy#4763]
Once our forward-auth-like features are done (https://github.com/caddyserver/caddy/pull/4739), we should make sure it works with Tailscale, similar to https://tailscale.com/blog/tailscale-auth-nginx/.

**Comment by francislavoie** [comment]
We got confirmation that it does work from a user: https://twitter.com/coldmacchiato_/status/1524683154237407232

I opened a PR to make some minor improvements in response to that #4783

**Comment by prez** [comment]
Would it be perhaps be possible to have a directive similar to `basicauth` (`tailscaleauth`?) that just contains a path to the tailscale socket, removing the boilerplate?
I'm not very informed, so disregard this if the "boilerplate" is actually needed. But it would certainly be awesome for users to have a simple directive to accomplish this. I would love to replace my usage of `basicauth` with it.

**Comment by francislavoie** [comment]
@prez the `forward_auth` directive _is_ the shortcut. For Tailscale, it would look like this, once #4783 is merged (which includes an improvement to `copy_headers` to support renaming headers when copying)

```
forward_auth unix//run/tailscale.nginx-auth.sock {
	uri /auth
	header_up Remote-Addr {remote_host}
	header_up Original-URI {uri}
	copy_headers {
		Tailscale-User>X-Webauth-User
		Tailscale-Name>X-Webauth-Name
		Tailscale-Login>X-Webauth-Login
		Tailscale-Tailnet>X-Webauth-Tailnet
		Tailscale-Profile-Picture>X-Webauth-Profile-Picture
	}
}
```

Ideally, those two `header_up` lines would not be needed, since they're passing redundant information to Tailscale; we already send the same info in `X-Forwarded-For` and `X-Forwarded-Uri`, but Tailscale doesn't read from those headers right now. I'll ask Tailscale to add support for that.

Either way, it's not a blocker, and you can use this right now with v2.5.1 if you don't need `copy_headers` header renaming and your upstream app can read from `Tailscale-*` headers. In which case, it would look like this, for v2.5.1:

```
forward_auth unix//run/tailscale.nginx-auth.sock {
	uri /auth
	header_up Remote-Addr {remote_host}
	header_up Original-URI {uri}
	copy_headers Tailscale-User Tailscale-Name Tailscale-Login Tailscale-Tailnet Tailscale-Profile-Picture
}
```

**Comment by francislavoie** [comment]
I opened an issue https://github.com/tailscale/tailscale/issues/4680 to see if Tailscale can smooth this over a bit.

**Comment by prez** [comment]
@francislavoie 

Thank you for the explanation. But this is mainly for backend applications that are under your control - is my understanding correct? 
I was looking for a `basicauth` replacement, that makes the route accessible only if the incoming connection is a tailscale one, to use with existing directives.

So for example, to use with `file_server`: instead of 
```
route /files*
    uri strip_prefix files
    basicauth / {
        user passhash
    }
    file_server dir
}
```
we would have something like
```
route /files*
    uri strip_prefix files
    tailscale_auth args
    file_server dir
}
```
or, for example, with a `reverse_proxy`:
```
route /socket* {
    reverse_proxy @websockets http://127.0.0.1:8081
    tailscale_auth args
}
```

To summarise, something where `uri /path` is not needed (as it's already inside a route block) and possibly the headers aren't either. I'm not sure if this is possible, but it would be awesome.

**Comment by francislavoie** [comment]
That's exactly what forward_auth does though. https://tailscale.com/blog/tailscale-auth-nginx/
